[Block Checkout] Hide shipping toggle when only local pickup methods exist

mikejolley opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

On checkout, if local pickup is enabled we show a toggle between delivery and collection:

![Screenshot 2025-01-07 at 15 08 34](https://github.com/user-attachments/assets/51e16254-b65d-4d01-b38a-098865528b65)

If local pickup is disabled, this toggle is hidden. This however was not the case if delivery was disabled but local pickup was enabled. This PR fixes this.

Now, if you have only local pickup configured, there will be no toggle visible. Uses the existing `wc_get_shipping_method_count` core functionality to determine if shipping methods are configured or not.

This PR fixes some styling issues in the pickup area also so the spacing matches payment methods.

Closes #48509

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to WooCommerce > Settings > Shipping
2. Remove all shipping methods and zones
3. Go to WooCommerce > Settings > Shipping > Local PIckup. Enable it and setup some pickup locations.
4. As a customer, add something to cart and go to checkout. Confirm you only see local pickup methods and there is no delivery/collection toggle.
5. Go to WooCommerce > Settings > Shipping
6. Add a shipping method to any zone and save.
7. As a customer, add something to cart and go to checkout. Confirm there is a delivery/collection toggle and it functions accordingly.
8. Go to WooCommerce > Settings > Shipping. 
9. Toggle the "enabled" checkbox off and save.
10. As a customer, add something to cart and go to checkout. Confirm you only see local pickup methods and there is no delivery/collection toggle.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

mikejolley committed:
Adjust pickup location styling to match payment methods

mikejolley committed:
Spacing

mikejolley committed:
Hide shipping/delivery switcher if SHIPPING_METHODS_EXIST is false

mikejolley committed:
shippingMethodsExist should factor in legacy

mikejolley committed:
changelog

mikejolley committed:
Put the readmore component back

opr reviewed (changes requested):
Thanks for working on this Mike, I noticed there's an failing test related to shipping, not sure what would cause that, the changes don't seem like they'd affect that. Maybe the first arg to `wc_get_shipping_method_count` that you changed is causing an issue, but that shouldn't affect store API routes 🤔 

> Automattic\WooCommerce\Tests\Blocks\StoreApi\Utilities\OrderControllerTests::test_validate_selected_shipping_methods
Automattic\WooCommerce\StoreApi\Exceptions\RouteException: Sorry, this order requires a shipping option.
>
> /var/www/html/wp-content/plugins/woocommerce/src/StoreApi/Utilities/OrderController.php:549
>/var/www/html/wp-content/plugins/woocommerce/tests/php/src/Blocks/StoreApi/Utilities/OrderControllerTests.php:41

This isn't failing locally either, seems like the setup in CI has some different shipping methods somehow.

I think we should write tests for this behaviour, especially around the `Local pickup is included with legacy shipping methods since they do not support shipping zones.` logic, seems like this could be easily broken by changes elsewhere in the codebase.

Besides this PHP fail, it's testing well for me following the instructions.

mikejolley committed:
Fix wc_get_shipping_method_count logic breakage due to params

mikejolley committed:
Remove debug code

mikejolley committed:
Add test and support for 3rd party methods

mikejolley committed:
Reset shipping when cart resets to prevent stale data in the shipping class (this was breaking tests)

mikejolley commented:
- Added an e2e test for the toggle visibility
- Fixed a core bug with how rate counts are cached
- Fixed a core bug with lingering packages after cart is cleared
- Fix test suite tear downs to clean up cart after running tests

opr commented on plugins/woocommerce/includes/class-wc-cart.php:
```diff
@@ -651,6 +651,7 @@ public function empty_cart( $clear_persistent_cart = true ) {
 		}
 
 		$this->fees_api->remove_all_fees();
+		WC()->shipping()->reset_shipping();
```
Do we need to update any docs anywhere for this? I doubt we documented this in so much detail but could be worth checking?

mikejolley replied:
I've added a little docblock. I don't see other docs for it.

opr replied:
Looks good, thanks!

opr reviewed (approved):
Working well now, followed the instructions and I placed an order with a virtual product and a normal product, both were fine 👌🏼

mikejolley committed:
Add since docblock
